hail/src/main/scala/is/hail/expr/types/physical/PTuple.scala

```diff
@@ -5,6 +5,10 @@ import is.hail.expr.ir.EmitMethodBuilder
 import is.hail.expr.types.TTuple
 import is.hail.utils._
 
+object PTuple {
+  def apply(args: PType*): PTuple = PTuple(args.toFastIndexedSeq, required = false)
```

Comment:
Take required as the first argument like we do in the TTuple constructor.

Reply:
I'm pretty sure this is going to get cleaned up soon anyway -- and remember that we plan to flip the requiredness so that it only exists on fields / array elements and not on every ptype.

If you want me to fix this, I will, though

Reply:
Ah, hmm.

Will we have `case class OptionalPType(pt: PType, required: Boolean)` or something similar?  I'm thinking about the type of something like `(ArrayRef a i)` where `a`'s elements are required.  In this case, do we want to generate code that knows/takes advantage of the fact result of the `ArrayRef` must exist?

Reply:
C/C++ doesn't really make a distinction between, on the value side, between objects that are represented on the stack vs heap.  Both are just memory.  So `A *ap = new A(args)` and `void f() { A a(args); ... }` represent an instance of `A` in the same way in two locations.  The result of an `ArrayRef` is a (stack-allocated) temporary.  The question is, as an object, how is a `pt: PType`-typed value represented?  It is potentially missing, so it needs a missing bit.  So I think it makes sense to have a ... discussion completed in person.